### PR TITLE
CDAP-6114 Virtual Machine documentation inconsistent

### PR DESCRIPTION
**Pull request by awholegunch**
Passes [Quick Build 5](http://builds.cask.co/browse/CDAP-DQB32-5)

Updates the Virtual Machine and Docker startup instructions.

Pages of interest:
- [Virtual Machine](http://builds.cask.co/artifact/CDAP-DQB32/shared/build-5/Docs-HTML/3.5.0-SNAPSHOT/en/developers-manual/getting-started/standalone/virtual-machine.html)
- [Docker Image](http://builds.cask.co/artifact/CDAP-DQB32/shared/build-5/Docs-HTML/3.5.0-SNAPSHOT/en/developers-manual/getting-started/standalone/docker.html)

Fix for https://issues.cask.co/browse/CDAP-6114

### CODE REVIEW

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 1]
```diff
@@ -178,8 +178,42 @@ Docker and CDAP Applications
...
-   :start-line: 4
+.. tabbed-parsed-literal::
+   :tabs: "Linux Virtual Machine"
+   :independent:
+
+   $ mvn archetype:generate \
+       -DarchetypeGroupId=co.cask.cdap \
+       -DarchetypeArtifactId=cdap-app-archetype \
+       -DarchetypeVersion=\ |release|
+
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Why is this in the Docker doc?

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 2]
```diff
@@ -178,8 +178,42 @@ Docker and CDAP Applications
...
+.. tabbed-parsed-literal::
+   :tabs: "Linux Virtual Machine"
+   :independent:
+
+   $ cd /opt/cdap/sdk
+   $ ./bin/cdap.sh start
+     . . .
+   $ ./bin/cdap.sh stop
+
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Again, why is this in the Docker doc? FYI, the VM starts the SDK automatically, too.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 3]
```diff
@@ -178,8 +178,42 @@ Docker and CDAP Applications
...
+Note that starting CDAP is not necessary if you use the Docker image, as it
+starts the Standalone CDAP automatically on startup.
+
+Once CDAP is started successfully, in a web browser you will be able to see the CDAP
+UI running at ``http://localhost:9999``, where you can deploy example applications and
+interact with CDAP. 
+
+Note that in the case of the Docker image, you will need to substitute 
+the Docker VM's IP address for ``localhost`` in the web browser address bar.
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Incorrect. If you have started the container as we've specified, localhost works as the Docker daemon proxies the ports. Docker is documented from the standard docker daemon, not Mac or Windows-based wrapper solutions, which launch in a VM. If you intend to document that usage, it needs to be documented separately.

File: cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst [hunk 4]
```diff
@@ -20,26 +22,58 @@ The CDAP Standalone Virtual Machine is configured with the recommended settings
 
 - 4 GB of RAM
 - Ubuntu Desktop Linux
-- 10 GB of disk space
+- 40 GB of virtual disk space
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst:26]: Why? The image is 4.3GB only.

File: cdap-docs/developers-manual/source/getting-started/start-stop-cdap.rst [hunk 5]
```diff
@@ -33,4 +33,4 @@ UI running at ``http://localhost:9999``, where you can deploy example applicatio
 interact with CDAP. 
 
 Note that in the case of the Docker image, you will need to substitute 
-the Docker VM's IP address for ``localhost`` in the web browser address bar.
\ No newline at end of file
+the Docker VM's IP address for ``localhost`` in the web browser address bar.
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/start-stop-cdap.rst]: This is again only true for wrapper-based usage. Normal docker usage, with a docker daemon running containers on the localhost will proxy the ports on `0.0.0.0` so localhost definitely will work. If you're running your docker container within a VM, you'll need to replace the IP.

File: cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst:26]: Well, it wasn't 10GB either. As we were giving the configuration, I changed it to the actual configuration I found. It's probably a sparse or an expanding drive, taking up as much space as it actually requires.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 1 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Each of the three formats (Zip, VM, Docker) has a single page dedicated to them describing the steps needed to start developing. I have for Docker (like I did for the VM) replaced what was there with more specific instructions: removed references to Windows in the starting and stoping of CDAP, etc.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: This is noted on the following line, but we can move it up. These are here so that people know how to stop (and restart) the SDK as required, especially as they need to if they are following some examples.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I will download the Docker image and run it (I haven't been able to run it for a long time) and check these instructions.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 1 above)
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Change the tab label to `Docker container` then. :-)

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Except that isn't how it works in the container image, at all. The container runs the standalone. If it stops, the process exits and the container exits. There is no starting and stopping or restarting it. You stop or restart the container, instead. The container is the standalone. It's not like a virtual machine.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 1 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Actually, I think your first question (Why is this here?) is more germane...I've run the Docker on Mac, will inflict it on my wife's Windows machine, and a linux box, and update everything accordingly...I see what you mean about how different all the approaches are, depending on the host OS.

**Comment by awholegunch** [comment]
On-hold as I have major revisions to come on this as I re-run Docker on the different platforms.

File: cdap-docs/developers-manual/source/getting-started/start-stop-cdap.rst (hunk 5 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/start-stop-cdap.rst]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 6]
```diff
@@ -49,7 +49,7 @@ Kitematic and then download, start, and connect to an instance of CDAP.
 
 #. Select the desired version.
    Note that the tag **latest** is the last version that was put up at Docker Hub, which is not the 
-   necessarily the most-current version:
+   necessarily the most-current version, which is |literal-version|:
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Let's change "most-current" to "desired" here, since this doc may be referenced after additional releases are made. A user reading the 3.5.8 document should be pointed to the 3.5.8 (or whatever... example) tag of the Docker image from Docker Hub.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 7]
```diff
@@ -70,11 +70,10 @@ Kitematic and then download, start, and connect to an instance of CDAP.
         :width: 8in
         :align: center
 
-
 #. This shows that the CDAP instance is listening on the internal port ``9999`` within the
    Docker instance, while the Docker instance exposes that port on the external IP address
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: s/instance/container/
s/instance/host/
s/exposes/proxies/
s/external/Virtual Machine/

A Docker container EXPOSEs ports, but when they're available outside the container, they've been proxied from the Docker host. In this case, the Docker host is the virtual machine.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 8]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
 ==========================
 
-- Docker is available for a variety of platforms. Download and install Docker in your environment by
-  following the `platform-specific installation instructions <https://docs.docker.com/engine/installation/>`__
-  from `Docker.com <https://docker.com>`__ to verify that Docker is working and has
-  started correctly.
-  
-  If you are not running on Linux, you need to start the Docker Virtual Machine (VM) before you
-  can use containers. For example:
+Docker is available for a variety of platforms. Download and install Docker in your
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Ehh... this isn't really true. Docker runs on Linux. It requires the Linux container capabilities in the Linux kernel.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 9]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
 Docker from a Command Line
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:23]: This should be first. This is _THE_ way to run Docker and really the _only_ way you'll usually see documented.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 10]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
-  to determine the Docker VM's IP address. You will need to use that address as the host
-  name when either connecting to the CDAP UI or making an HTTP request.
+   .. tabbed-parsed-literal::
+     :tabs: "Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+     
+     $ docker-machine create --driver virtualbox cdap
+     $ docker-machine env cdap
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:41]: Are these some `boot2docker` commands?

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 11]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
+   .. tabbed-parsed-literal::
+     :tabs: "Linux or Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+ 
+     $ docker pull caskdata/cdap-standalone:|release|
 
-- Start the *Docker CDAP Virtual Machine* with:
+#. Start the *Docker CDAP Virtual Machine* with:
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Start the CDAP Standalone Docker container

It's not a virtual machine.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 12]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
-    :mapping: linux,windows
-    :dependent: linux-windows
-    :languages: console,shell-session
+   .. tabbed-parsed-literal::
+     :tabs: "Linux or Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+ 
+     $ docker run -t -i -p 9999:9999 -p 10000:10000 caskdata/cdap-standalone:|release|
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Let's change this:

`docker run -d --name cdap-standalone -p 9999:9999 -p 10000:10000 caskdata/cdap-standalone:|release|`

This will name the container `cdap-standalone` which we can reference below, versus having the Docker daemon auto-generating a name.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 13]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
+#. To control the CDAP instance, use this command, substituting one of ``start``, ``restart``,
+   or ``stop`` for ``<command>``:
+
+   .. tabbed-parsed-literal::
+     :tabs: "Linux or Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+     
+     $ docker exec -d $(docker ps | grep cdap-standalone | awk '{print $1}') /opt/cdap/sdk/bin/cdap.sh <command>
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Use `docker exec -d cdap-standalone /opt/cdap/sdk/bin/cdap.sh <command>` now that we've given the container a known name.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 14]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
-  at ``http://localhost:9999``. Replace ``localhost`` with the Docker VM's IP address 
-  (such as ``192.168.59.103``) that you obtained earlier. Start a browser and enter the
-  address to access the CDAP UI.
+   .. tabbed-parsed-literal::
+     :tabs: "Linux or Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+     
+     $ docker exec -d $(docker ps | grep cdap-standalone | awk '{print $1}') /opt/cdap/sdk/bin/cdap.sh stop
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Use `docker exec -d cdap-standalone /opt/cdap/sdk/bin/cdap.sh stop` instead... naming the container makes it so much easier, no?

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 15]
```diff
@@ -86,87 +85,97 @@ Kitematic and then download, start, and connect to an instance of CDAP.
...
+     $ docker-machine stop cdap
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:129]: This command is probably some `boot2docker` thing and doesn't exist on a normal Docker installation.

File: cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst [hunk 16]
```diff
@@ -20,26 +22,58 @@ The CDAP Standalone Virtual Machine is configured with the recommended settings
...
+Use the ``cdap.sh`` script to start and stop the Standalone CDAP:
+
+.. tabbed-parsed-literal::
+   :tabs: "Linux Virtual Machine"
+   :independent:
+
+   $ cd /opt/cdap/sdk
+   $ ./bin/cdap.sh start
+     . . .
+   $ ./bin/cdap.sh stop
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst]: The SDK actually has an init script in the VM, which is how it's started. We should probably use the same in our documentation, or document both methods. In any case, the SDK runs as root, so you must sudo.

```
sudo /etc/init.d/cdap-sdk start
sudo /etc/init.d/cdap-sdk stop
```

or

```
sudo /opt/cdap/sdk/bin/cdap.sh start
sudo /opt/cdap/sdk/bin/cdap.sh stop
```

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 6 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 7 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 8 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Yes, Docker kind of skates over this detail by referring to it as a "Docker environment".

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 11 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 12 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Changed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 13 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Changed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 15 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:129]: It's used on OS X and Windows...though I haven't tried it yet on Windows. I can add an extra tab to move it to only those tabs.

Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:129]: Actually, I take that back. It's for all platforms. See https://docs.docker.com/machine/reference/stop/

File: cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst (hunk 16 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/virtual-machine.rst]: I tried this on Oracle VirtualBox, using CDAP 3.4.1. I didn't need `sudo` to stop or start cdap. Sure this is required?

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 8 above)
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fair enough.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 15 above)
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:129]: But there's no VM on Linux... I'm pretty sure that binary doesn't even exist on my Linux install. You can get a `docker-base` cluster from Coopr for testing.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 9 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:23]: Opps. missed this one. Will address it.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 10 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:41]: No, these are `docker-machine` commands, which (I think) replaced `boot2docker`.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 9 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:23]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 17]
```diff
@@ -11,12 +11,129 @@ Docker Image
 
 
 Docker is one of the easiest ways to start working with CDAP without having to manually
-configure anything. A Docker image with the CDAP SDK pre-installed is available on the Docker Hub
+configure anything. A Docker container with the CDAP SDK pre-installed is available on the Docker Hub
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: This was correct before.

The thing that we create and ship to Docker Hub is an image. That image is taken and instantiated as a container. So... running == container, the thing you build/pull is an image.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 18]
```diff
@@ -11,12 +11,129 @@ Docker Image
...
+   .. tabbed-parsed-literal::
+     :tabs: "Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+     
+     $ docker-machine create --driver virtualbox cdap
+     $ docker-machine env cdap
+     
+   This will create a new Docker virtual machine named ``cdap`` and print out the environment.
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: "This will create a new virtual machine using VirtualBox named cdap and print out the environment."

Docker runs inside the virtual machine.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 19]
```diff
@@ -11,12 +11,129 @@ Docker Image
...
+     :dependent: linux-windows
+     :languages: console,shell-session
+     
+     .. Linux or Mac OS X
+     
+     $ docker -d cdap-standalone /opt/cdap/sdk/bin/cdap.sh <command>
+
+     .. Windows
+     
+     > docker -d cdap-standalone /opt/cdap/sdk/bin/cdap.sh <command>
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: This was probably my fault... but it's "docker exec -d ..."

**Comment by wolf31o2** [comment]
A couple more minor comments... otherwise 👍 LGTM

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 17 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 18 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: True, but that's not how Docker's documentation describes it. They describe it (at https://docs.docker.com/engine/installation/windows/) as:

`<snip>`
   [3]. Create a new Docker VM.

```
   docker-machine create --driver virtualbox my-default
```

   The command also creates a machine configuration in the
   C:\USERS\USERNAME.docker\machine\machines directory. You only need to run the create
   command once. Then, you can use docker-machine to start, stop, query, and otherwise manage
   the VM from the command line.

`</snip>`

but what about using:

```
This will create a new Docker virtual machine using VirtualBox named ``cdap`` and print out the environment.
```

so that way it is both accurate and matches up with their way of describing it.

Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I like `This will create a new virtual machine using VirtualBox for Docker named cdap and print out...` but I'm not married to it. What you have is fine, really.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 20]
```diff
@@ -14,9 +14,128 @@ Docker is one of the easiest ways to start working with CDAP without having to m
...
+
+#. Start the *CDAP Standalone Docker container* with:
+
+   .. tabbed-parsed-literal::
+     :tabs: "Linux or Mac OS X",Windows
+     :mapping: linux,windows
+     :dependent: linux-windows
+     :languages: console,shell-session
+ 
+     $ docker run --detach --name cdap-standalone -p 9999:9999 -p 10000:10000 caskdata/cdap-standalone:|release|
```
Comment (wolf31o2) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I'd like to keep the docs as close as possible to the docs seen on Docker Hub, which invariably use `-d` versus the long form. Also, less typing means less typos.

**Comment by wolf31o2** [comment]
One comment. If you just change those, then 👍 LGTM
